### PR TITLE
Demo rollout via MaskAngle

### LINKED ISSUE #23721: Verify Instrument 2.0 Python Interface
Changes recently made #23706 would be best verified by trying to apply Instrument 2.0 to refactor an existing python algorithm.

First `ag` search yields `MaskAngle` so attempt to roll out to that. 

Getting this right will pave the way for future developers to modernise our geometry access on the python side.

### PR DESCRIPTION
**Pull request by OwenArnold**
fixes #23721 

`SpectrumInfo` and `ComponentInfo` applied to MaskAngle. No use of Instrument 1.0 at all.
This change should not affect release notes.

**Tester**

Code review, verify tests pass, run `MaskAngle` and check everything looks OK.

### CODE REVIEW

File: Framework/PythonInterface/inc/MantidPythonInterface/kernel/Policies/VectorToNumpy.h [hunk 1]
```diff
@@ -111,12 +110,11 @@ template <typename VectorType> struct VectorToNumpyImpl {
...
 };
 
-template <typename T>
-struct VectorToNumpy_Requires_StdVector_Return_By_Value {};
+template <typename T> struct VectorToNumpy_Requires_StdVector_Return {};
 } // namespace
 
 /**
- * Implements a return value policy that
+ * Implements a return value policy that clones
```
Comment (SimonHeybrock) [inline Framework/PythonInterface/inc/MantidPythonInterface/kernel/Policies/VectorToNumpy.h]: This sentence has one word too many now.

File: Framework/PythonInterface/inc/MantidPythonInterface/kernel/Policies/VectorToNumpy.h [hunk 2]
```diff
@@ -87,8 +87,7 @@ template <typename ConversionPolicy> struct VectorRefToNumpy {
         typename std::remove_reference<T>::type>::type;
     // MPL compile-time check that T is a reference to a std::vector
     using type = typename boost::mpl::if_c<
-        boost::mpl::and_<std::is_reference<T>,
```
Comment (SimonHeybrock) [inline Framework/PythonInterface/inc/MantidPythonInterface/kernel/Policies/VectorToNumpy.h]: I am not sure I understand. The class name `VectorRefToNumpy` still implies that is checks for reference? I also have not convinced myself that this doesn't break any existing behavior, will do so once this is clarified.

File: Framework/PythonInterface/mantid/geometry/src/Exports/DetectorInfo.cpp [hunk 3]
```diff
@@ -90,5 +90,9 @@ void export_DetectorInfo() {
 
       .def("rotation", rotation, (arg("self"), arg("index")),
            "Returns the absolute rotation of the detector where the detector "
-           "is identified by 'index'.");
+           "is identified by 'index'.")
+      .def("detectorIDs", &DetectorInfo::detectorIDs,
+           return_value_policy<
+               Mantid::PythonInterface::Policies::VectorToNumpy>(),
```
Comment (SimonHeybrock) [inline Framework/PythonInterface/mantid/geometry/src/Exports/DetectorInfo.cpp]: Why is this not exposed as a read-only numpy array like, e.g., `MatrixWorkspace::readX`?

File: Framework/PythonInterface/plugins/algorithms/MaskAngle.py [hunk 4]
```diff
@@ -42,61 +44,56 @@ def PyInit(self):
...
+        else:
+            hasInstrument = len(ws.componentInfo()) > 0
+        if not hasInstrument:
             issues["Workspace"] = "Workspace must have an associated instrument."
-
         return issues
 
+    def _get_phi(self, det_index, detector_info):
+        pos = detector_info.position(det_index)
+        # see Detector.cpp for phi definition
```
Comment (SimonHeybrock) [inline Framework/PythonInterface/plugins/algorithms/MaskAngle.py]: Note that that phi definition is wrong in general, which is the reason we do not have it in `DetectorInfo`: It assumes that the beam is aligned with Z (which is true for most instruments currently, as far as I know, but not for those used in unit tests, I think). Basically this "phi" is the phi in a spherical coordinate system, but in neutron scattering I understand it is usually understood as the azimuth of the scattering process.

File: Framework/PythonInterface/plugins/algorithms/MaskAngle.py [hunk 5]
```diff
@@ -42,61 +44,56 @@ def PyInit(self):
...
-                    if tt>= ttmin and tt<= ttmax:
-                        detlist.append(det.getID())
-
-        if len(detlist)> 0:
-            mantid.simpleapi.MaskDetectors(Workspace=ws,DetectorList=detlist)
-        else:
+        detectorInfo = ws.detectorInfo()
+        det_ids = detectorInfo.detectorIDs()
+        masked_ids = list()
+        for i in range(numspec):
```
Comment (SimonHeybrock) [inline Framework/PythonInterface/plugins/algorithms/MaskAngle.py]: Can't you use the `SpectrumInfo` iterator here?

File: Framework/PythonInterface/plugins/algorithms/MaskAngle.py [hunk 6]
```diff
@@ -42,61 +44,56 @@ def PyInit(self):
...
+            if not spectrumInfo.isMonitor(i):
+                # Get the first detector of spectrum. Ignore time aspects.
+                det_index = spectrumInfo.getSpectrumDefinition(i)[0][0]
+                if angle_phi:
+                    val = self._get_phi(det_index, detectorInfo)
+                else:
+                    # Two theta
+                    val =detectorInfo.twoTheta(det_index)
```
Comment (SimonHeybrock) [inline Framework/PythonInterface/plugins/algorithms/MaskAngle.py]: The original code used positions and angles of the spectrum, why is this changed to use the first detector?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
Questions and comments:

File: Framework/PythonInterface/mantid/geometry/src/Exports/DetectorInfo.cpp (hunk 3 above)
Comment (OwenArnold) [reply Framework/PythonInterface/mantid/geometry/src/Exports/DetectorInfo.cpp]: Agreed will change this

File: Framework/PythonInterface/plugins/algorithms/MaskAngle.py (hunk 5 above)
Comment (OwenArnold) [reply Framework/PythonInterface/plugins/algorithms/MaskAngle.py]: Wasn't available when I started writing this, yes will update

File: Framework/PythonInterface/plugins/algorithms/MaskAngle.py (hunk 6 above)
Comment (OwenArnold) [reply Framework/PythonInterface/plugins/algorithms/MaskAngle.py]: My mistake. Will correct. Probably the cause of the current workflow algorithm test failure.

File: Framework/PythonInterface/inc/MantidPythonInterface/kernel/Policies/VectorToNumpy.h (hunk 2 above)
Comment (OwenArnold) [reply Framework/PythonInterface/inc/MantidPythonInterface/kernel/Policies/VectorToNumpy.h]: This is about the return type as specified by function being exposed.

File: Framework/PythonInterface/plugins/algorithms/MaskAngle.py (hunk 4 above)
Comment (SimonHeybrock) [reply Framework/PythonInterface/plugins/algorithms/MaskAngle.py]: Maybe add a comment here that this is strictly speaking a bug?

**Comment by SimonHeybrock** [comment]
After the last change, in C++ the iterators will not work anymore for `const` workspaces. Since this is not required right now I think we can postpone fixing this.

**Review by SimonHeybrock — DISMISSED** [review]

File: Framework/PythonInterface/plugins/algorithms/MaskAngle.py [hunk 7]
```diff
@@ -48,61 +50,58 @@ def PyInit(self):
...
-
         return issues
 
+    def _get_phi(self, spectra_pos):
+        '''
+        The implementation here assumes that z is the beam direction.
+        That assumption is not universally true, it depends on the geometry configuration.
+        This returns the phi spherical coordinate value
+        '''
+        return math.fabs(math.atan2(spectra_pos.Y(), spectra_pos.X()))
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/MaskAngle.py]: Did we finally came up with a solution to get the phi angle in a "universal" way? I do not want to implement the same function everywhere where I need phi in Instrument 2.0

Comment (AndreiSavici) [reply Framework/PythonInterface/plugins/algorithms/MaskAngle.py]: Also, no need to import math. You already imported numpy

File: Framework/API/inc/MantidAPI/SpectrumInfoIterator.h [hunk 8]
```diff
@@ -69,18 +70,18 @@ class MANTID_API_DLL SpectrumInfoIterator
...
-  bool equal(const SpectrumInfoIterator &other) const {
+  bool equal(const SpectrumInfoIterator<T> &other) const {
     return getIndex() == other.getIndex();
   }
 
-  const SpectrumInfoItem &dereference() const { return m_item; }
+  SpectrumInfoItem<T> &dereference() const { return m_item; }
 
-  SpectrumInfoItem m_item;
+  mutable SpectrumInfoItem<T> m_item;
```
Comment (SimonHeybrock) [inline Framework/API/inc/MantidAPI/SpectrumInfoIterator.h]: I do not like the use of `mutable` here, but I am also a bit confused: In `SpectrumInfoItem` you have a `T *m_spectrumInfo`, and a non-const method `setMasked`. Either this should not compile, or it lets you call `setMasked()` even in the `const` version. Am I missing something here?

Comment (OwenArnold) [reply Framework/API/inc/MantidAPI/SpectrumInfoIterator.h]: You instantiate what you use, you would get a compile time error if you tried to call `setMasked` on the non-const version. That might not be intuitive in the way that it would be expressed as a compiler message to the client. One option would be to use something like: ```static_assert(!std::is_const<T>::value, "Operation disabled on const T");``` to give a bit more context to compiler errors for this.

Comment (OwenArnold) [reply Framework/API/inc/MantidAPI/SpectrumInfoIterator.h]: The reason that the mutable is required is to get things working with the `iterator_facade`, that specifies a `dereference` function required on the derived class as:

```
SpectrumInfoItem<T> &dereference() const;
```
If you have a better way around this would be interested to get your ideas.

Comment (SimonHeybrock) [reply Framework/API/inc/MantidAPI/SpectrumInfoIterator.h]: I suppose it is ok for now.

**Review by SimonHeybrock — COMMENTED** [review]
@AndreiSavici Regarding getting phi in Instrument-2.0: Are we happy with changing the behavior of the old `getDetector().getPhi()` as well? If not, we still need a clear and distiguishable name for it in Instrument-2.0, and I do not have a suggestion, unfortunately.

**Comment by SimonHeybrock** [comment]
@AndreiSavici Regarinding support getting phi in Instrument-2.0: Are we happy with changing the behavior of the old `getDetector().getPhi()` as well? If not, we still need a clear and distiguishable name for it in Instrument-2.0, and I do not have a suggestion, unfortunately.

File: Framework/PythonInterface/plugins/algorithms/MaskAngle.py (hunk 7 above)
Comment (AndreiSavici) [reply Framework/PythonInterface/plugins/algorithms/MaskAngle.py]: We can probably come up with some reasonable default parameter names and values, so when calling `getPhi` with no parameters you get the current behavior.

**Comment by AndreiSavici** [comment]
We can probably come up with some reasonable default parameter names and values, so when calling `getPhi` with no parameters you get the current behavior.

**Review by SimonHeybrock — DISMISSED** [review]

**Review by SimonHeybrock — COMMENTED** [review]
My original review was somehow marked as stale by github...

**Review by SimonHeybrock — APPROVED** [review]

**Comment by OwenArnold** [comment]
@mantidproject/gatekeepers This PR is awaiting a final review.

**Comment by martyngigg** [comment]
Apparently GitHub now dismisses reviews if something new is pushed after the review has been completed.
